collector/ethtool_linux_test.go

```diff
@@ -0,0 +1,125 @@
+// Copyright 2021 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package collector
+
+import (
+	"bufio"
+	"fmt"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"syscall"
+	"testing"
+
+	"github.com/go-kit/kit/log"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+type EthtoolFixture struct {
+	fixturePath string
+}
+
+func (e *EthtoolFixture) Stats(intf string) (map[string]uint64, error) {
+	res := make(map[string]uint64)
+
+	fixtureFile, err := os.Open(filepath.Join(e.fixturePath, intf))
+	if e, ok := err.(*os.PathError); ok && e.Err == syscall.ENOENT {
+		// The fixture for this interface doesn't exist. That's OK because it replicates
+		// an interface that doesn't support ethtool.
+		return res, nil
+	}
+	if err != nil {
+		return res, err
+	}
+	defer fixtureFile.Close()
+
+	scanner := bufio.NewScanner(fixtureFile)
+	for scanner.Scan() {
+		line := scanner.Text()
+		if strings.HasPrefix(line, "#") {
+			continue
+		}
+		if strings.HasPrefix(line, "NIC statistics:") {
+			continue
+		}
+		line = strings.Trim(line, " ")
+		items := strings.Split(line, ": ")
+		val, err := strconv.ParseUint(items[1], 10, 64)
+		if err != nil {
+			return res, err
+		}
+		res[items[0]] = val
+	}
+
+	return res, err
+}
+
+func NewEthtoolTestCollector(logger log.Logger) (Collector, error) {
+	collector, err := makeEthtoolCollector(logger)
+	collector.stats = &EthtoolFixture{
+		fixturePath: "fixtures/ethtool/",
+	}
+	if err != nil {
+		return nil, err
+	}
+	return collector, nil
+}
+
+func TestEthtoolCollector(t *testing.T) {
+	testcases := []string{
+		prometheus.NewDesc("node_ethtool_align_errors", "Network interface align_errors", []string{"device"}, nil).String(),
```

Comment:
Most of these seem to be counters, right? In that case they should have a _total suffix

Reply:
This one is an example of a driver-specific metric (in this case [r8152](https://github.com/torvalds/linux/blob/9d31d2338950293ec19d9b095fbaa9030899dcb4/drivers/net/usb/r8152.c#L781)). The kernel doesn't tell us if it's a counter or a gauge. Some "well known" metrics are pre-populated in makeEthtoolCollector() and do have the _total suffix. Later when we get support for standardized metrics in kernel 5.13+, we'll be in a position to name and type them appropriately.

Reply:
Ah got it, yeah if it's driver specific we shouldn't bother 'translating' it.